graylog2-server/src/main/java/org/graylog2/events/ClusterEventService.java

```diff
@@ -0,0 +1,203 @@
+                        final EventBus serverEventBus,
+                        final EventBus clusterEventBus) {
+        this.nodeId = checkNotNull(nodeId);
+        this.dbCollection = checkNotNull(dbCollection);
+        this.objectMapper = checkNotNull(objectMapper);
+        this.serverEventBus = checkNotNull(serverEventBus);
+        this.clusterEventBus = checkNotNull(clusterEventBus);
+
+        this.clusterEventBus.register(this);
+    }
+
+    @VisibleForTesting
+    static DBCollection prepareCollection(final MongoConnection mongoConnection) {
+        final DB db = mongoConnection.getDatabase();
+
+        DBCollection coll;
+        if (!db.collectionExists(COLLECTION_NAME)) {
+            final DBObject options = BasicDBObjectBuilder.start()
+                    .add("capped", true)
+                    .add("size", 32768)
+                    .get();
+
+            try {
+                coll = db.createCollection(COLLECTION_NAME, options);
+            } catch (MongoException e) {
+                // Error code 48 == "collection already exists"
+                if(e.getCode() == 48) {
+                    coll = db.getCollection(COLLECTION_NAME);
+                } else {
+                    throw e;
+                }
+            }
+        } else {
+            coll = db.getCollection(COLLECTION_NAME);
+        }
+
+        if (!db.getCollection(COLLECTION_NAME).isCapped()) {
+            LOG.warn("The MongoDB collection \"{}\" should be capped but isn't. "
+                    + "Please drop the collection and restart Graylog", COLLECTION_NAME);
+        }
+
+        coll.createIndex(DBSort.asc("producer"));
+        coll.createIndex(DBSort.asc("consumers"));
+        coll.addOption(Bytes.QUERYOPTION_TAILABLE | Bytes.QUERYOPTION_AWAITDATA);
+
+        return coll;
+    }
+
+    private DBCursor<ClusterEvent> eventCursor(NodeId nodeId) {
+        // Resorting to ugly MongoDB Java Client because of https://github.com/devbliss/mongojack/issues/88
+        final DBObject producerClause = new BasicDBObject("producer", new BasicDBObject("$ne", nodeId.toString()));
+        final BasicDBList consumersList = new BasicDBList();
+        consumersList.add(nodeId.toString());
+        final DBObject consumersClause = new BasicDBObject("consumers", new BasicDBObject("$nin", consumersList));
+        final BasicDBList and = new BasicDBList();
+        and.add(producerClause);
+        and.add(consumersClause);
+        final DBObject query = new BasicDBObject("$and", and);
+
+        return dbCollection.find(query)
+                .sort(DBSort.asc("$natural"))
+                .addOption(Bytes.QUERYOPTION_TAILABLE)
+                .addOption(Bytes.QUERYOPTION_AWAITDATA);
+    }
+
+    private void updateConsumers(final String eventId, final NodeId nodeId) {
+        final WriteResult<ClusterEvent, String> writeResult = dbCollection.updateById(eventId, DBUpdate.addToSet("consumers", nodeId.toString()));
```

Comment:
What is the WriteConcern we are using here?
If I'm not mistaken, it will default to NORMAL which means that only network errors lead to exceptions, but not server errors.
We would not be able to detect any failure to update the collection (neither here, not anywhere else).
I think at the least it should use ACKNOWLEDGED or even MAJORITY.